loopd/daemon.go

```diff
@@ -159,100 +172,221 @@ func (d *Daemon) Run() error {
+		// up its connection now, nothing else needs to be shut down at
+		// this point.
+		clientCleanup()
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
 
 	// Start the swap client itself.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, d.statusChan)
+		err := d.impl.Run(d.mainCtx, d.statusChan)
 		if err != nil {
-			log.Error(err)
+			// Notify the main error handler goroutine that
+			// we exited unexpectedly here. We don't have to
+			// worry about blocking as the internal error
+			// channel is sufficiently buffered.
+			d.internalErrChan <- err
 		}
 		log.Infof("Swap client stopped")
-
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
-
-		cancel()
 	}()
 
 	// Start a goroutine that broadcasts swap updates to clients.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Waiting for updates")
-		d.processStatusUpdates(mainCtx)
+		d.processStatusUpdates(d.mainCtx)
 	}()
 
-	// Start the grpc server.
-	wg.Add(1)
+	// Last, start our internal error handler. This will return exactly one
+	// error or nil on the main error channel to inform the caller that
+	// something went wrong or that shutdown is complete. We don't add to
+	// the wait group here because this goroutine will itself wait for the
+	// stop to complete and signal its completion through the main error
+	// channel.
 	go func() {
-		defer wg.Done()
-
-		log.Infof("RPC server listening on %s", grpcListener.Addr())
-
-		if restListener != nil {
-			log.Infof("REST proxy listening on %s", restListener.Addr())
+		var runtimeErr error
+
+		// There are only two ways this goroutine can exit. Either there
+		// is an internal error or the caller requests shutdown. In both
+		// cases we wait for the stop to complete before we signal the
+		// caller that we're done.
+		select {
+		case runtimeErr = <-d.internalErrChan:
+			log.Errorf("Runtime error in daemon, shutting down: "+
+				"%v", runtimeErr)
+
+		case <-d.quit:
 		}
 
-		err = grpcServer.Serve(grpcListener)
-		if err != nil {
-			log.Error(err)
-		}
+		// We need to shutdown before sending the error on the channel,
+		// otherwise a caller might exit the process too early.
+		d.stop()
+		log.Info("Daemon exited")
+
+		// The caller expects exactly one message. So we send the error
+		// even if it's nil because we cleanly shut down.
+		d.ErrChan <- runtimeErr
 	}()
 
-	interruptChannel := make(chan os.Signal, 1)
-	signal.Notify(interruptChannel, os.Interrupt)
+	return nil
+}
 
-	// Run until the users terminates loopd or an error occurred.
-	select {
-	case <-interruptChannel:
-		log.Infof("Received SIGINT (Ctrl+C).")
+// Stop tries to gracefully shut down the daemon. A caller needs to wait for a
+// message on the main error channel indicating that the shutdown is completed.
```

Comment:
Why don't we wait for the error channel in this function? So we stop+block waiting for our error, then the called knows we're done once this function unblocks.